feat(order): 주문 기능 구현(고객/사장)

3uomlkh opened the pull request:
## 변경 사항
 - 장바구니와 메뉴 연결
 - 주문 기능 구현

## 구현 내용
- 고객
   - 주문을 생성하는 기능 구현
   - 주문을 취소하는 기능 구현
   - 주문 시 가게 영업 시간과 최소 주문 금액을 충족하지 않을 시 주문 실패
- 사장
   - 들어온 주문을 수락하는 기능
   - 들어온 주문을 거절하는 기능 

## 추가해야하는 사항 
- `Menu` 도메인과 `Store` 도메인 연동 필요
- 세션 방식에서 JWT로 변경 필요
- 사장님이 주문 현황을 더 다양하게 바꿀 수 있도록 수정 필요
- 로그 구현 필요

3uomlkh committed:
refactor(Cart): 장바구니와 메뉴 연결 구현

### 장바구니와 메뉴 연결
- 기존에 임시 필드로 사용하던 menuId 삭제 후, Menu와 연결
- 현재 메뉴와 가게가 연결되어 있지 않아 가게 이름은 가져오지 못함
- 같은 가게 메뉴만 장바구니에 담는 조건도 추후 구현 예정

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>

3uomlkh committed:
Merge branch 'develop' into feat/order

3uomlkh committed:
Merge branch 'develop' into feat/order

3uomlkh committed:
feat(Order): 주문 생성 기능 구현 #12

### 주문 생성 기능 구현
- 장바구니에 CartEntity 리스트를 반환하는 메서드 추가
- 장바구니가 비었을 시 주문 실패
- 현재 시각이 가게 영업 시간이 아닐 시 주문 실패
- 총 주문 금액이 최소 주문 금액 미만일 때 주문 실패
- 주문 저장 후, 상세 내용을 주문 리스트에 저장
- 주문 완료되면 장바구니 비워지고 'PENDING' 상태로 전환
- 메뉴와 가게 연결 후 수정 필요

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>

3uomlkh committed:
feat(Order): 주문 취소 기능 구현(고객) #12

### 주문 취소 기능 구현
- 주문이 'PENDING' 상태일 경우에만 삭제되도록 구현
- 본인의 주문이 아닐 경우, 주문이 존재하지 않는 경우 예외 발생
- OrderStatus에 CANCELED 상태 추가

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>

3uomlkh committed:
feat(Order): 주문 수락 및 거절 기능 구현(사장) #12

### 주문 수락 및 거절 기능 구현
- 주문이 'PENDING' 상태일 경우에만 수락 및 거절 되도록 구현
- 본인의 가게 주문만 수락/거절할 수 있도록 구현
- OrderStatus를 ACCEPTED 혹은 REJECTED로 변경
- 추후 '음식 조리중', '배달중', '배달완료' 상태 변경이 가능하도록 구현 예정

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>

mmj-159 commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/cart/service/UserCartService.java:
```diff
@@ -31,12 +34,15 @@ public AddCartResponseDto addCartItem(
                 .orElseThrow(() -> new InvalidRequestException(ErrorCartMessage.USER_NOT_FOUND.getMessage()));
 
+        MenuEntity menu = menuRepository.findById(dto.getMenuId())
+                .orElseThrow(() -> new InvalidRequestException("메뉴가 존재하지 않습니다."));
```
() -> new InvalidRequestException(MenuErrorMessage.NOT_FOUND_MENU.getErrorMessage())
에러 메시지는 이렇게 하시면 될 것 같은데, 다른 생각있으신가요??

mixedsider replied:
다른 도메인에서 Repository를 불러오는 것은 권장하지 않고 싶습니다.
되도록 서비스 레이어를 불러와주세요.

3uomlkh replied:
넵 수정하겠습니다.

mmj-159 reviewed (approved):
reivew : 코드 검토 완료
MenuErrorMessage 이 부분만 수정해주시면 될 것 같습니다!

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/cart/entity/CartEntity.java:
```diff
@@ -21,18 +22,16 @@ public class CartEntity extends BaseEntity {
     @JoinColumn(name = "user_id", nullable = false)
```
1대 1관계여서 관계를 적어두시지 않은건가요?

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/order/entity/UserOrderListEntity.java:
```diff
@@ -0,0 +1,34 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "menu_id", nullable = false)
+    MenuEntity menu;
```
private 빠져있습니다.

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/order/service/OwnerOrderService.java:
```diff
@@ -0,0 +1,48 @@
+    public void acceptOrder(Long ownerId, Long orderId) {
+        UserOrderEntity order = userOrderService.getOrderById(orderId);
+
+        if (!ownerId.equals(order.getStore().getUser().getId())) {
```
ObjectUtils.nullSafeEquals() 를 사용하는게 
더 NullSafe한 코딩이 될 것같습니다.

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/order/service/OwnerOrderService.java:
```diff
@@ -0,0 +1,48 @@
+    public void rejectOrder(Long ownerId, Long orderId) {
+        UserOrderEntity order = userOrderService.getOrderById(orderId);
+
+        if (!ownerId.equals(order.getStore().getUser().getId())) {
```
위와 같습니다.

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/order/service/OwnerOrderService.java:
```diff
@@ -0,0 +1,48 @@
+        UserOrderEntity order = userOrderService.getOrderById(orderId);
+
+        if (!ownerId.equals(order.getStore().getUser().getId())) {
+            throw new InvalidRequestException("본인의 가게 주문만 수락할 수 있습니다.");
```
메세지 Enum 타입으로 안하신것에 이유가 있을까요?

3uomlkh committed:
feat(Order): 주문 에러 메세지 Enum으로 수정 #12

### 에러 메세지 수정
- ErrorOrderMessage enum 클래스 생성
- ErrorCartMessage 수정
- Service 쪽 에러 메세지 전체 수정

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/cart/service/UserCartService.java:
```diff
@@ -22,21 +26,25 @@ public class UserCartService {
 
     private final CartRepository cartRepository;
     private final UserRepository userRepository;
+    private final MenuRepository menuRepository;
```
되도록 서비스 레이어를 사용해주세요.

mixedsider commented on src/main/java/xyz/tomorrowlearncamp/outsourcing/domain/order/service/UserOrderService.java:
```diff
@@ -0,0 +1,81 @@
+        UserOrderEntity order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new InvalidRequestException(ErrorOrderMessage.ORDER_NOT_FOUND.getMessage()));
+
+        if (!order.getUser().getId().equals(userId)) {
```
위에있는 nullSafe 코딩과 같습니다.

mixedsider reviewed (approved):
review : 코드 검토 완료

- 전체적으로 잘 짜신것같습니다.
- 다만 조금더 NullSafe 한 코딩을 신경을 써주셨으면 하는 바람
- 다른 서비스 레이어에서 다른 도메인의 서비스 레이어를 불러오는 것이 좀더 바람직하다는 점을 말씀 드리고 싶습니다.

3uomlkh committed:
refactor(Order): 피드백 반영 수정 #12

### 피드백 반영
- private 접근 제한자 추가
- UserCartService에서 userRepository를 사용하지 않도록 수정
- equals() 비교 시 nullsafe하도록 수정
- MenuRepository에 MenuEntity를 반환하는 메서드가 없어 이후 수정 예정

Signed-off-by: 3uomlkh <3uomlkh@gmail.com>
